frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java

```diff
@@ -0,0 +1,111 @@
+public abstract class BaseRenameAction extends BaseGitMacheteRepositoryReadyAction
+    implements
+      IBranchNameProvider,
+      IExpectsKeyGitMacheteRepository {
+
+  @Override
+  @UIEffect
+  protected void onUpdate(AnActionEvent anActionEvent) {
+    super.onUpdate(anActionEvent);
+
+    val presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    val branchName = getNameOfBranchUnderAction(anActionEvent);
+    val branch = branchName != null
+        ? getManagedBranchByName(anActionEvent, branchName)
+        : null;
+
+    if (branch == null) {
+      presentation.setEnabled(false);
+      presentation.setDescription(getNonHtmlString("action.GitMachete.description.disabled.undefined.machete-branch")
+          .fmt("Rename", getQuotedStringOrCurrent(branchName)));
+    } else {
+      presentation.setDescription(
+          getNonHtmlString("action.GitMachete.BaseRenameAction.description").fmt(branch.getName()));
+    }
+  }
+
+  @Override
+  @ContinuesInBackground
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    val project = getProject(anActionEvent);
+    val gitRepository = getSelectedGitRepository(anActionEvent);
+    val currentBranchName = getNameOfBranchUnderAction(anActionEvent);
+    val branchLayout = getBranchLayout(anActionEvent);
+    val branchLayoutWriter = RuntimeBinding.instantiateSoleImplementingClass(IBranchLayoutWriter.class);
+
+    if (gitRepository == null || currentBranchName == null || branchLayout == null) {
+      return;
+    }
+
+    rename(project, gitRepository, currentBranchName, branchLayout, branchLayoutWriter);
+  }
+
+  @ContinuesInBackground
+  @UIEffect
+  private void rename(Project project,
+      GitRepository gitRepository,
+      String currentBranchName, BranchLayout branchLayout, IBranchLayoutWriter branchLayoutWriter) {
+    val gitNewBranchDialog = new GitNewBranchDialog(project, Collections.singletonList(gitRepository),
+        getString("action.GitMachete.BaseRenameAction.description").fmt(currentBranchName),
+        currentBranchName,
+        /* showCheckOutOption */ false,
+        /* showResetOption */ false,
+        /* showSetTrackingOption */ false,
+        /* localConflictsAllowed */ false,
+        GitBranchOperationType.RENAME);
+
+    val options = gitNewBranchDialog.showAndGetOptions();
+
+    if (options != null) {
+      Path macheteFilePath = gitRepository.getMacheteFilePath();
+      val newBranchLayout = branchLayout.rename(currentBranchName, options.getName());
+
+      runWriteActionOnUIThread(() -> {
+        MacheteFileWriter.writeBranchLayout(
+            macheteFilePath,
+            branchLayoutWriter,
+            newBranchLayout,
+            /* backupOldLayout */ true,
+            /* requestor */ this);
+
+        GitBrancher.getInstance(project).renameBranch(currentBranchName, options.getName(),
```

Comment:
Huh unfortunately... I'm getting a race condition on the newest version built from this PR:

<img width="1678" alt="image" src="https://user-images.githubusercontent.com/3383210/213446683-012bc0c1-3e89-410d-b571-7eb1208fe3c6.png">

Pls analyze the flow once more... I can help using my machine if you can't reproduce it locally

Reply:
I cannot reproduce it - for me it's 
<img width="386" alt="image" src="https://user-images.githubusercontent.com/19799111/213674784-e9f03fcb-7790-485d-8ac1-daee1b7c51a5.png">
now.

The version with the redundant backgroundable wroked better Xd reverting the revert

Reply:
wow. I think that we need both the backgroundable and the `waitForCreationOfLocalBranch`. why?

1. the backgroundable wraps the rename of entry nad of branch 
2. `waitForCreationOfLocalBranch` keeps the backgroundable alive and blocks the other backgroundables, e.g. `GitMacheteRepositoryUpdateBackgroundable` - a backgroundable responsible for the notification about not existing branch ! 

Together it seems to work like some atomic operation in the scope of backgroundables 😆 

probably no the best approach, but it's working and it's here already

Reply:
> keeps the backgroundable alive and blocks the other backgroundables

Hmmm but aren't these backgroundables somewhat independent of each other? 🤔 as in, it's **not** like UI thread where there's just one thread, and indeed only task can be running at any given moment. There are multiple threads designated for handling background tasks.

> probably no the best approach, but it's working and it's here already

Still, if it works, then let's go for that... lemme check on my machine

Reply:
I thought so too - but it works like that